# CSCI-473: Applied Data Science
**Homework Assignment 1 - Cleaning and Summarizing Palmer Penguins Data**

## Student Information
**Alhassane Diallo**  
004001562
alhassane.diallo@bison.howard.edu  
CSCI-473  
Homework Assignment 1 - Cleaning and Summarizing Palmer Penguins Data

## Introduction

This assignment uses the Palmer Penguins dataset, which contains information on 344 penguins across three species and three islands. The purpose of this assignment is to practice using pandas to inspect and clean the data, handle missing values, correct data types, validate measurements, summarize findings, and export the cleaned dataset.

## Task 1: Load and Audit the Dataset

In [1]:
import pandas as pd 
df = pd.read_csv("penguins_raw.csv") ##load the file 

Load the provided CSV file into a pandas DataFrame named df.
Your program must display:
1. the first five rows;

In [2]:
##read first 5 rows
df.head(5)

,studyName,Sample Number,Species,Region,Island,Stage,Individual ID,Clutch Completion,Date Egg,Culmen Length (mm),Culmen Depth (mm),Flipper Length (mm),Body Mass (g),Sex,Delta 15 N (o/oo),Delta 13 C (o/oo),Comments
0,PAL0708,1,Adelie Penguin (Pygoscelis adeliae),Anvers,Torgersen,"Adult, 1 Egg Stage",N1A1,Yes,2007-11-11,39.1,18.7,181.0,3750.0,MALE,NaN,NaN,Not enough blood for isotopes.
1,PAL0708,2,Adelie Penguin (Pygoscelis adeliae),Anvers,Torgersen,"Adult, 1 Egg Stage",N1A2,Yes,2007-11-11,39.5,17.4,186.0,3800.0,FEMALE,8.94956,-24.69454,NaN
2,PAL0708,3,Adelie Penguin (Pygoscelis adeliae),Anvers,Torgersen,"Adult, 1 Egg Stage",N2A1,Yes,2007-11-16,40.3,18.0,195.0,3250.0,FEMALE,8.36821,-25.33302,NaN
3,PAL0708,4,Adelie Penguin (Pygoscelis adeliae),Anvers,Torgersen,"Adult, 1 Egg Stage",N2A2,Yes,2007-11-16,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Adult not sampled.
4,PAL0708,5,Adelie Penguin (Pygoscelis adeliae),Anvers,Torgersen,"Adult, 1 Egg Stage",N3A1,Yes,2007-11-16,36.7,19.3,193.0,3450.0,FEMALE,8.76651,-25.32426,NaN


2. the last five rows

In [3]:
## read last 5 rows 
df.tail(5)

,studyName,Sample Number,Species,Region,Island,Stage,Individual ID,Clutch Completion,Date Egg,Culmen Length (mm),Culmen Depth (mm),Flipper Length (mm),Body Mass (g),Sex,Delta 15 N (o/oo),Delta 13 C (o/oo),Comments
339,PAL0910,64,Chinstrap penguin (Pygoscelis antarctica),Anvers,Dream,"Adult, 1 Egg Stage",N98A2,Yes,2009-11-19,55.8,19.8,207.0,4000.0,MALE,9.70465,-24.53494,NaN
340,PAL0910,65,Chinstrap penguin (Pygoscelis antarctica),Anvers,Dream,"Adult, 1 Egg Stage",N99A1,No,2009-11-21,43.5,18.1,202.0,3400.0,FEMALE,9.37608,-24.40753,Nest never observed with full clutch.
341,PAL0910,66,Chinstrap penguin (Pygoscelis antarctica),Anvers,Dream,"Adult, 1 Egg Stage",N99A2,No,2009-11-21,49.6,18.2,193.0,3775.0,MALE,9.46180,-24.70615,Nest never observed with full clutch.
342,PAL0910,67,Chinstrap penguin (Pygoscelis antarctica),Anvers,Dream,"Adult, 1 Egg Stage",N100A1,Yes,2009-11-21,50.8,19.0,210.0,4100.0,MALE,9.98044,-24.68741,NaN
343,PAL0910,68,Chinstrap penguin (Pygoscelis antarctica),Anvers,Dream,"Adult, 1 Egg Stage",N100A2,Yes,2009-11-21,50.2,18.7,198.0,3775.0,FEMALE,9.39305,-24.25255,NaN


3. the number of rows and columns


In [4]:
## the number of rows and columns
df.shape

(344, 17)

4. all column names

In [5]:
## all column names
df.columns

Index(['studyName', 'Sample Number', 'Species', 'Region', 'Island', 'Stage',
       'Individual ID', 'Clutch Completion', 'Date Egg', 'Culmen Length (mm)',
       'Culmen Depth (mm)', 'Flipper Length (mm)', 'Body Mass (g)', 'Sex',
       'Delta 15 N (o/oo)', 'Delta 13 C (o/oo)', 'Comments'],
      dtype='str')

5. the data type of every column;

In [6]:
## the data type of every column
df.dtypes

studyName                  str
Sample Number            int64
Species                    str
Region                     str
Island                     str
Stage                      str
Individual ID              str
Clutch Completion          str
Date Egg                   str
Culmen Length (mm)     float64
Culmen Depth (mm)      float64
Flipper Length (mm)    float64
Body Mass (g)          float64
Sex                        str
Delta 15 N (o/oo)      float64
Delta 13 C (o/oo)      float64
Comments                   str
dtype: object

6. the output of info();

In [7]:
## the output of info()
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 17 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   studyName            344 non-null    str    
 1   Sample Number        344 non-null    int64  
 2   Species              344 non-null    str    
 3   Region               344 non-null    str    
 4   Island               344 non-null    str    
 5   Stage                344 non-null    str    
 6   Individual ID        344 non-null    str    
 7   Clutch Completion    344 non-null    str    
 8   Date Egg             344 non-null    str    
 9   Culmen Length (mm)   342 non-null    float64
 10  Culmen Depth (mm)    342 non-null    float64
 11  Flipper Length (mm)  342 non-null    float64
 12  Body Mass (g)        342 non-null    float64
 13  Sex                  333 non-null    str    
 14  Delta 15 N (o/oo)    330 non-null    float64
 15  Delta 13 C (o/oo)    331 non-null    float64
 16  C

7. descriptive statistics for the numerical columns; and

In [8]:
## descriptive statistics for the numerical columns; and
df.describe()

,Sample Number,Culmen Length (mm),Culmen Depth (mm),Flipper Length (mm),Body Mass (g),Delta 15 N (o/oo),Delta 13 C (o/oo)
count,344.000000,342.000000,342.000000,342.000000,342.000000,330.000000,331.000000
mean,63.151163,43.921930,17.151170,200.915205,4201.754386,8.733382,-25.686292
std,40.430199,5.459584,1.974793,14.061714,801.954536,0.551770,0.793961
min,1.000000,32.100000,13.100000,172.000000,2700.000000,7.632200,-27.018540
25%,29.000000,39.225000,15.600000,190.000000,3550.000000,8.299890,-26.320305
50%,58.000000,44.450000,17.300000,197.000000,4050.000000,8.652405,-25.833520
75%,95.250000,48.500000,18.700000,213.000000,4750.000000,9.172123,-25.062050
max,152.000000,59.600000,21.500000,231.000000,6300.000000,10.025440,-23.787670


8. the number of missing values in each column.

In [9]:
## the number of missing values in each column.
df.isna().sum()

studyName                0
Sample Number            0
Species                  0
Region                   0
Island                   0
Stage                    0
Individual ID            0
Clutch Completion        0
Date Egg                 0
Culmen Length (mm)       2
Culmen Depth (mm)        2
Flipper Length (mm)      2
Body Mass (g)            2
Sex                     11
Delta 15 N (o/oo)       14
Delta 13 C (o/oo)       13
Comments               290
dtype: int64

### Final eval

In task one we loaded into the data file and began inspecting it for data quality issuses and concers using functions such ass .isna(), .info(), .dtypes, .head(), and .shape()  

The audit revealed that data was missing which can indicate that some penguins wasn't measured properly and that some data may not be reliable. Missing data like length, depth or mass can scew the means but becuase the precentage is pretty low, I would come to the conclusion that the effect in this scenerio isnt dignificant.

    - penguins missing sex or delta concentration most likley didn't have or have enough blood taken
    - data set includes penguins that werent measured at all 

## Task 2: Clean Column Names and Correct Data Types

Create a working copy of the original DataFrame before cleaning:

In [10]:
## create copy of data file 
raw_df = df.copy()
clean_df = df.copy()

Rename the columns so that they use clear snake case names.

In [11]:
## rename files 
clean_df = clean_df.rename(columns = {
    "studyName": "study_name",
    "Sample Number": "sample_number",
    "Species" : "species",
    "Island" : "island",
    "Clutch Completion" : "clutch_completion",
    "Date Egg": "date_egg",
    "Individual ID" : "individual_id",
    "Culmen Length (mm)" : "culmen_length_mm",
    "Culmen Depth (mm)" : "culmen_depth_mm",
    "Flipper Length (mm)" : "flipper_length_mm",
    "Body Mass (g)" : "body_mass_g",
    "Sex" : "sex",
    "Region": "region",
    "Stage": "stage",
    "Delta 15 N (o/oo)": "delta_15_n",
    "Delta 13 C (o/oo)": "delta_13_c",
    "Comments": "comments"
})

1. convert date egg to a pandas date-time data type. Use pd.to datetime() and set conversion
    errors to "coerce";

In [12]:
## convert date to pandas data type
clean_df ["date_egg"] = pd.to_datetime(
    clean_df["date_egg"],
    errors = "coerce" ## if can't convert, don't raise error
                                    )

2. use pd.to numeric() with errors="coerce" on the four primary measurement columns to
    ensure that they are numerical; and

In [13]:
## convert length to numeric data type 
clean_df ["culmen_length_mm"] = pd.to_numeric( 
    clean_df["culmen_length_mm"],
        errors = "coerce"
    )

In [14]:
## convert depth to numeric data type 
clean_df ["culmen_depth_mm"] = pd.to_numeric( 
    clean_df["culmen_depth_mm"],
        errors = "coerce" )

In [15]:
## convert flipper to numeric data type 
clean_df ["flipper_length_mm"] = pd.to_numeric( 
    clean_df["flipper_length_mm"],
        errors = "coerce" )

In [16]:
## convert mass to numeric data type 
clean_df ["body_mass_g"] = pd.to_numeric( 
    clean_df["body_mass_g"],
        errors = "coerce" )

4. print the data types after conversion.
    The four primary measurement columns are:
   culmen length mm,
   culmen depth mm,
   flipper length mm, and
   body mass g.

In [17]:
clean_df[["culmen_length_mm", "culmen_depth_mm", "flipper_length_mm", "body_mass_g"]].dtypes

culmen_length_mm     float64
culmen_depth_mm      float64
flipper_length_mm    float64
body_mass_g          float64
dtype: object

### Final Report

Report Requirement  
Describe the name and type corrections you made. Include a screenshot showing the cleaned column
names and resulting data types.

    I did both a column name correction (Snake Case) and  a data type correction using .rename(), to_datetime() and to_numeric(). By doing this we are able to standerdize the data types acorss each colomun and this will come in handy later when interpertation the data.

In [18]:
clean_df.dtypes

study_name                      str
sample_number                 int64
species                         str
region                          str
island                          str
stage                           str
individual_id                   str
clutch_completion               str
date_egg             datetime64[us]
culmen_length_mm            float64
culmen_depth_mm             float64
flipper_length_mm           float64
body_mass_g                 float64
sex                             str
delta_15_n                  float64
delta_13_c                  float64
comments                        str
dtype: object

## Task 3: Clean Categories, Missing Values, and Duplicates

#### A. Inspect and standardize categories

In [19]:
## tells us how many unique species there are and how many exist 
clean_df["species"].value_counts(dropna=False)

species
Adelie Penguin (Pygoscelis adeliae)          152
Gentoo penguin (Pygoscelis papua)            124
Chinstrap penguin (Pygoscelis antarctica)     68
Name: count, dtype: int64

In [20]:
## tells us how many unique islands there are and how many exist in each column
clean_df["island"].value_counts(dropna=False)

island
Biscoe       168
Dream        124
Torgersen     52
Name: count, dtype: int64

In [21]:
clean_df["clutch_completion"].value_counts(dropna=False)

clutch_completion
Yes    308
No      36
Name: count, dtype: int64

In [22]:
clean_df["sex"].value_counts(dropna=False)

sex
MALE      168
FEMALE    165
NaN        11
Name: count, dtype: int64

In [23]:
## Replace any placeholder value in sex, such as ".", with pd.NA.
clean_df["sex"]= clean_df["sex"].replace(".", pd.NA)

In [24]:
## Standardize the species values to the shorter category names: Adelie, Chinstrap, and Gentoo.
clean_df["species"] = clean_df["species"].replace ({"Adelie Penguin (Pygoscelis adeliae)": "Adelie",
    "Chinstrap penguin (Pygoscelis antarctica)": "Chinstrap",
    "Gentoo penguin (Pygoscelis papua)": "Gentoo"
})

#### B. Check duplicates

Report:  
1. the number of exact duplicate rows; 

In [25]:
## number of duplicate rows
clean_df.duplicated().sum()

np.int64(0)

There are 0 duplicated rows

2. all occurrences of repeated individual id values. Use duplicated() with the subset argument set to the identifier column and keep=False.

In [26]:
## total amount of occurrences of repeated individual id values
clean_df.duplicated(subset = "individual_id", keep= False).sum()

np.int64(268)

There are 268 occurences of repeated I'd numbers

In [27]:
clean_df[clean_df.duplicated(subset="individual_id", keep=False)][
    ["individual_id", "species", "date_egg"]
    ].sort_values("individual_id")

,individual_id,species,date_egg
20,N11A1,Adelie,2007-11-12
198,N11A1,Gentoo,2008-11-09
21,N11A2,Adelie,2007-11-12
199,N11A2,Gentoo,2008-11-09
200,N12A1,Gentoo,2008-11-02
...,...,...,...
196,N8A1,Gentoo,2008-11-03
236,N8A1,Gentoo,2009-11-22
15,N8A2,Adelie,2007-11-16
237,N8A2,Gentoo,2009-11-22


    Report:  
    A duplicated row indicates a data entry error where as a repeated indentifiers suggest that the subject was visited multiple times and can indicate remeasurment, retest and that scientist revisted the field sit, possibly over mutliple seasons.

#### C. Handle missing values
    For the final analysis dataset, keep only rows that contain recorded values for all of the following
    required fields:
        • species,
        • island,
        • culmen length mm,
        • culmen depth mm,
        • flipper length mm,
        • body mass g, and
        • sex.

In [28]:
print("Numbers of rows before cleaning:", clean_df.shape[0])

Numbers of rows before cleaning: 344


In [29]:
## removeany row that has a missing value that falls under one of these primary columns
required_columns = [
    "species",
    "island",
    "culmen_length_mm",
    "culmen_depth_mm",
    "flipper_length_mm",
    "body_mass_g",
    "sex"
    ]
clean_df = clean_df.dropna(subset=required_columns).copy()

In [30]:
print("Numbers of rows After cleaning:", clean_df.shape[0])

Numbers of rows After cleaning: 333


#### Report

I used .value_counts(dropna=False) on species, island, clutch_completion, and sex to inspect all unique entries and identify unstandardized text or nonstandard missing placeholders. I replaced place holder values in sex by replacing them with pd.NA and I standerized species names to a simplified short name using .replace(). Then I evaluated exact duplicate rows across all columns using clean_df.duplicated().sum(), which returned 0 exact duplicates. Finally we used .dropna with a subset to drop rows with missing measurments that could effect final calculations 

## Task 4: Validate the Cleaned Data

Perform the following validation checks:
            
Use Boolean conditions and at least three assertions. Do not invent biological upper limits for
measurements unless a documented rule is provided. A large value is not automatically invalid.

Print a confirmation message after all assertions pass.  

Report Requirement
List the rules you validated and explain what each assertion checks. Include a screenshot showing the successful validation output.


 rule 1. Primary Measurement Values Must Be Positive
     verifies that the four primary measurement columns contain only positive values;

In [31]:
primary_measurements = [
    "culmen_length_mm",
    "culmen_depth_mm",
    "flipper_length_mm",
    "body_mass_g"
]

try:
    assert (clean_df[primary_measurements] > 0).all().all(), "Measurements are not all positive."
    print("Assertion passed: All primary measurements are positive.")
except AssertionError as e:
    print(f"Assertion failed: {e}")

Assertion passed: All primary measurements are positive.


Rule 2. Allowed Species Categories
    verifies that species column contains only Adelie, Chinstrap, and Gentoo;

In [32]:
allowed_species = ["Adelie", "Chinstrap", "Gentoo"]

try: 
    assert clean_df["species"].unique().isin(allowed_species).all(), "Species recorded is not allowed"
    print ("Assertion passed: All Recorded Species are allowed")
except AssertionError as e: 
    print(f"Assertion failed: {e}")    

Assertion passed: All Recorded Species are allowed


Rule 3: Allowed Island Categories
    verifies that island contains only the categories observed in the original dataset;

In [33]:
allowed_island= ['Torgersen', 'Biscoe', 'Dream']

try:
    assert clean_df ["island"].unique().isin(allowed_island).all(), "Contains island that doesnt belong in data set"
    print ("Island Column contains only the categories observed in the original dataset")
except AssertionError as e:
    print(f"Assertion failed: {e}")

Island Column contains only the categories observed in the original dataset


Rule 4: No Missing Values in Required Fields
    verifies that the required fields from Task 3 contain no missing values 


In [34]:
## check that required fields have no missing value
try: 
    assert clean_df[required_columns].notna().all().all(),"Dataset contains missing values"
    print("Data set contains no missing values")
except AssertionError as e: 
    print(f"Assertion failed:{e}")

Data set contains no missing values


Rule 5: No Exact Duplicate Rows
    verifies that the cleaned dataset contains no exact duplicate rows.

In [35]:
## verify no duplicates
try:
    assert clean_df.duplicated().sum() == 0,"Duplicate rows detected."
    print("Dataset contains no duplicated entries")
except AssertionError as e:
    print(f"Assertion failed:{e}")

Dataset contains no duplicated entries


### Report

For Task 4, I used try-except blocks and assert statements to ensure data quality and integrity before proceeding to analysis. I used (clean_df[primary_measurements] > 0).all().all() to verify that all physical measurements were strictly positive, .unique() and .isin() to confirm species and island names matched allowed categories, .notna().all().all() to check that no missing values remained in critical required fields, and .duplicated().sum() == 0 to confirm the removal of duplicate rows. All five assertions passed, confirming that the dataset contains valid categories, complete required records, and no duplicate or negative entries.

## Task 5: Filter, Sort, and Summarize the Data
Use the cleaned DataFrame to answer all of the following questions.

### A. Filtering and sorting

1. Display all Gentoo penguins with a body mass greater than 5,000 grams.

In [36]:
clean_df[(clean_df["species"] == "Gentoo") & (clean_df["body_mass_g"] > 5000)]

,study_name,sample_number,species,region,island,stage,individual_id,clutch_completion,date_egg,culmen_length_mm,culmen_depth_mm,flipper_length_mm,body_mass_g,sex,delta_15_n,delta_13_c,comments
153,PAL0708,2,Gentoo,Anvers,Biscoe,"Adult, 1 Egg Stage",N31A2,Yes,2007-11-27,50.0,16.3,230.0,5700.0,MALE,8.14756,-25.39369,NaN
155,PAL0708,4,Gentoo,Anvers,Biscoe,"Adult, 1 Egg Stage",N32A2,Yes,2007-11-27,50.0,15.2,218.0,5700.0,MALE,8.25540,-25.40075,NaN
156,PAL0708,5,Gentoo,Anvers,Biscoe,"Adult, 1 Egg Stage",N33A1,Yes,2007-11-18,47.6,14.5,215.0,5400.0,MALE,8.23450,-25.54456,NaN
159,PAL0708,8,Gentoo,Anvers,Biscoe,"Adult, 1 Egg Stage",N34A2,Yes,2007-11-27,46.7,15.3,219.0,5200.0,MALE,8.22673,-25.42760,NaN
161,PAL0708,10,Gentoo,Anvers,Biscoe,"Adult, 1 Egg Stage",N35A2,Yes,2007-11-27,46.8,15.4,215.0,5150.0,MALE,8.16310,-25.38017,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
267,PAL0910,116,Gentoo,Anvers,Biscoe,"Adult, 1 Egg Stage",N35A2,Yes,2009-11-25,55.1,16.0,230.0,5850.0,MALE,8.08354,-26.18161,NaN
269,PAL0910,118,Gentoo,Anvers,Biscoe,"Adult, 1 Egg Stage",N36A2,Yes,2009-12-01,48.8,16.2,222.0,6000.0,MALE,8.33825,-25.88547,NaN
273,PAL0910,122,Gentoo,Anvers,Biscoe,"Adult, 1 Egg Stage",N39A2,Yes,2009-11-22,50.4,15.7,222.0,5750.0,MALE,8.30166,-26.04117,NaN
274,PAL0910,123,Gentoo,Anvers,Biscoe,"Adult, 1 Egg Stage",N43A1,Yes,2009-11-22,45.2,14.8,212.0,5200.0,FEMALE,8.24246,-26.11969,NaN


2. Display all female Adelie penguins from Torgersen Island.

In [37]:
clean_df[(clean_df["species"]=="Adelie") & (clean_df["sex"]=="FEMALE") & (clean_df["island"]=="Torgersen")]

,study_name,sample_number,species,region,island,stage,individual_id,clutch_completion,date_egg,culmen_length_mm,culmen_depth_mm,flipper_length_mm,body_mass_g,sex,delta_15_n,delta_13_c,comments
1,PAL0708,2,Adelie,Anvers,Torgersen,"Adult, 1 Egg Stage",N1A2,Yes,2007-11-11,39.5,17.4,186.0,3800.0,FEMALE,8.94956,-24.69454,NaN
2,PAL0708,3,Adelie,Anvers,Torgersen,"Adult, 1 Egg Stage",N2A1,Yes,2007-11-16,40.3,18.0,195.0,3250.0,FEMALE,8.36821,-25.33302,NaN
4,PAL0708,5,Adelie,Anvers,Torgersen,"Adult, 1 Egg Stage",N3A1,Yes,2007-11-16,36.7,19.3,193.0,3450.0,FEMALE,8.76651,-25.32426,NaN
6,PAL0708,7,Adelie,Anvers,Torgersen,"Adult, 1 Egg Stage",N4A1,No,2007-11-15,38.9,17.8,181.0,3625.0,FEMALE,9.18718,-25.21799,Nest never observed with full clutch.
12,PAL0708,13,Adelie,Anvers,Torgersen,"Adult, 1 Egg Stage",N7A1,Yes,2007-11-15,41.1,17.6,182.0,3200.0,FEMALE,NaN,NaN,Not enough blood for isotopes.
15,PAL0708,16,Adelie,Anvers,Torgersen,"Adult, 1 Egg Stage",N8A2,Yes,2007-11-16,36.6,17.8,185.0,3700.0,FEMALE,NaN,NaN,Not enough blood for isotopes.
16,PAL0708,17,Adelie,Anvers,Torgersen,"Adult, 1 Egg Stage",N9A1,Yes,2007-11-12,38.7,19.0,195.0,3450.0,FEMALE,9.18528,-25.06691,NaN
18,PAL0708,19,Adelie,Anvers,Torgersen,"Adult, 1 Egg Stage",N10A1,Yes,2007-11-16,34.4,18.4,184.0,3325.0,FEMALE,8.47827,-25.23319,NaN
68,PAL0809,69,Adelie,Anvers,Torgersen,"Adult, 1 Egg Stage",N32A1,No,2008-11-11,35.9,16.6,190.0,3050.0,FEMALE,8.47781,-26.07821,Nest never observed with full clutch.
70,PAL0809,71,Adelie,Anvers,Torgersen,"Adult, 1 Egg Stage",N34A1,Yes,2008-11-14,33.5,19.0,190.0,3600.0,FEMALE,7.88863,-26.63085,NaN


3. Display the five penguins with the greatest flipper lengths. Show the species, island, sex, flipper length, and body mass.

In [38]:
clean_df[["species", "island", "sex", "flipper_length_mm", "body_mass_g"]].sort_values(
    by="flipper_length_mm", ascending=False).head(5)

,species,island,sex,flipper_length_mm,body_mass_g
215,Gentoo,Biscoe,MALE,231.0,5650.0
185,Gentoo,Biscoe,MALE,230.0,6050.0
265,Gentoo,Biscoe,MALE,230.0,5500.0
241,Gentoo,Biscoe,MALE,230.0,5550.0
153,Gentoo,Biscoe,MALE,230.0,5700.0


### B. Group summaries
    Calculate and display:
    Use groupby() and use agg() for the summaries requiring multiple statistics or columns. Round
    displayed means to two decimal places.

1. the number of penguins in each species;

In [39]:
clean_df.groupby("species").agg(
    penguin_size = ("species", "size")
)

,penguin_size
species,
Adelie,146
Chinstrap,68
Gentoo,119


Used groupby() to group penguins by species name, then used agg() with size to find the total in each group.  
The data shows Adelie is the most frequent species in this dataset, making up almost half of the total population sampled.

2. the mean body mass for each species;

In [40]:
clean_df.groupby("species").agg(
    penguin_mass = ("body_mass_g", "mean")
).round(2)

,penguin_mass
species,
Adelie,3706.16
Chinstrap,3733.09
Gentoo,5092.44


Used groupby() to group penguins by species name, then agg() with mean to find the mean of each group and added round() to round to the second decimal point.  
The data shows Gentoo penguins are significantly heavier than the other two species

3. the mean culmen length, mean flipper length, and mean body mass for each species; 

In [41]:
clean_df.groupby("species").agg(
mean_culmen_length=("culmen_length_mm", "mean"),
mean_flipper_length=("flipper_length_mm", "mean"),
mean_body_mass_g=("body_mass_g", "mean"),
).round(2)

,mean_culmen_length,mean_flipper_length,mean_body_mass_g
species,,,
Adelie,38.82,190.10,3706.16
Chinstrap,48.83,195.82,3733.09
Gentoo,47.57,217.24,5092.44


Used groupby() to group penguins by species name, then agg() with mean to calculate the mean culmen length, mean flipper length, and mean body mass for each species.  
The data shows that flipper length increases as mass increases.

4. the number and mean body mass of penguins grouped by both species and sex.


In [42]:
clean_df.groupby(["species","sex"]).agg(
    penguin_count = ("species","size"),
    mean_body_mass_g = ("body_mass_g", "mean"),
).round(2)

penguin_count  mean_body_mass_g
species   sex                                    
Adelie    FEMALE             73           3368.84
          MALE               73           4043.49
Chinstrap FEMALE             34           3527.21
          MALE               34           3938.97
Gentoo    FEMALE             58           4679.74
          MALE               61           5484.84

Used groupby(["species", "sex"]) to split the dataset by both species and sex, then used .agg with size and species to find to total count of penguins that fall under each sex for each of the different species. I then did the same but used mean to caluculate the mean body mass, ending with round() to round each mean to the 2 decimal point. \
The data shows to be well balanced between mall and female test, male seem to consistenly be heavier than there femal counterparts.


### Report 

For Task 5, I filtered, sorted, and summarized the dataset by combing Boolean conditions to filter specific subsets. I used groupby(), and agg() with mean and size to find both the mean and sample size  by species and I used sort_values(by="flipper_length_mm", ascending=False).head(5) to identify that the five penguins with the largest flipper lengths. I also added round(2) to round number values the second decimal point. The aggregations showed that Adelie is the most common species, Gentoos have the highest average body mass, and across all species, male penguins are consistently heavier than female penguins while maintaining balanced sample sizes.

## Task 6: Export and Verify the Cleaned Dataset
Save the final cleaned DataFrame as: palmer penguins clean.csv \
Do not save the DataFrame index as a separate column. Reload the saved file into a new DataFrame
and display: \
• its first five rows; \
• its shape; and \
• its missing-value counts for the required fields. \
Compare the reloaded shape with the final shape before saving. They should be the same.
\
\
Report Requirement \
Describe how you verified the exported file. Include a screenshot showing the reloaded data, its shape, and the missing-value check.


In [43]:
clean_df.to_csv("palmer_penguins_clean.csv", index=False)

In [44]:
checked_df=pd.read_csv("palmer_penguins_clean.csv")

In [45]:
checked_df.head(5)

,study_name,sample_number,species,region,island,stage,individual_id,clutch_completion,date_egg,culmen_length_mm,culmen_depth_mm,flipper_length_mm,body_mass_g,sex,delta_15_n,delta_13_c,comments
0,PAL0708,1,Adelie,Anvers,Torgersen,"Adult, 1 Egg Stage",N1A1,Yes,2007-11-11,39.1,18.7,181.0,3750.0,MALE,NaN,NaN,Not enough blood for isotopes.
1,PAL0708,2,Adelie,Anvers,Torgersen,"Adult, 1 Egg Stage",N1A2,Yes,2007-11-11,39.5,17.4,186.0,3800.0,FEMALE,8.94956,-24.69454,NaN
2,PAL0708,3,Adelie,Anvers,Torgersen,"Adult, 1 Egg Stage",N2A1,Yes,2007-11-16,40.3,18.0,195.0,3250.0,FEMALE,8.36821,-25.33302,NaN
3,PAL0708,5,Adelie,Anvers,Torgersen,"Adult, 1 Egg Stage",N3A1,Yes,2007-11-16,36.7,19.3,193.0,3450.0,FEMALE,8.76651,-25.32426,NaN
4,PAL0708,6,Adelie,Anvers,Torgersen,"Adult, 1 Egg Stage",N3A2,Yes,2007-11-16,39.3,20.6,190.0,3650.0,MALE,8.66496,-25.29805,NaN


In [46]:
checked_df[required_columns].isna().all().all()

np.False_

In [47]:
clean_df.shape

(333, 17)

In [48]:
checked_df.shape

(333, 17)

### Report 

For task 6, we saved the cleaned dataset to csv file and then we went ahead a used .isna() and .shape() to verify that no information was dropped from the dataset.  

The reloaded data (checked_df) frame has a shape 333 rows by 17 columns which macthes the shape of clean_df. This confirms that all 333 rows were preserved and that setting index=False successfully prevented an extra index column from being saved. I can also see that no values where dropped since the empty column checked returned all false.



## Cleaning Summary

1. Problems Found During the audit  
        I identified several data quality issues in the raw dataset: missing values in critical measurement columns (culmen_length_mm, culmen_depth_mm, flipper_length_mm, body_mass_g, and sex), non-standard column names containing spaces and special characters, inconsistent capitalization in categorical entries (such as sex), and potential duplicate records.  

2. How They Were Handled  
    To fix these issues, I standardized all column headers into clean, lower-case snake_case formatting. I converted categorical values like sex to uppercase for consistency across rows. For missing data, I used .dropna(subset=required_columns) to drop only the rows missing key biological measurements or species/island indicators while preserving records that only lacked optional metadata (such as isotope ratios). Finally, I checked for and removed duplicate entries using .drop_duplicates().  

3. Rows Removed  
        The original dataset contained 344 rows. Dropping rows with missing values across the core required columns removed 11 rows in total, bringing the final cleaned dataset (clean_df) to 333 rows.

4. Verification of Final Dataset
    I verified the final dataset using programmatic assertions wrapped in try-except blocks. I confirmed that all numerical measurements are strictly positive, species and island values belong exclusively to expected categories, no missing values remain in required_columns, and zero duplicate rows exist. Lastly, after exporting the data to palmer_penguins_clean.csv, I reloaded it as checked_df and verified that its shape (333, 17) perfectly matched clean_df with no unintended extra index columns or missing values.

## GEN Ai use statement 
Gen ai was used to  
1. help with erors
2. help understand code example and how different functions work


## Conclusion

This assignment thought me the different functions used when cleaning and validationg data and the process. It provided great practice and allowed me to get down to the basics, understanding how pandas work and things to check for such as missing values and duplicated columns. I also learned how to standerize naming columns and that problem can occur when data type isn't standerized across a column. As I finished the last task, I found myself looking over the first few task and updating my code to increase visibility, runtime and understanding with more complex code. 
